src/servers/src/http/loki.rs

```diff
@@ -282,6 +280,78 @@ async fn handle_pb_req(
     Ok(rows)
 }
 
+/// since we're hand-parsing the labels, if any error is encountered, we'll just skip the label
+/// note: pub here for bench usage
+/// ref:
+/// 1. encoding: https://github.com/grafana/alloy/blob/be34410b9e841cc0c37c153f9550d9086a304bca/internal/component/common/loki/client/batch.go#L114-L145
+/// 2. test data: https://github.com/grafana/loki/blob/a24ef7b206e0ca63ee74ca6ecb0a09b745cd2258/pkg/push/types_test.go
+pub fn parse_loki_labels(labels: &str) -> Result<BTreeMap<String, String>> {
+    let mut labels = labels.trim();
+    ensure!(
+        labels.len() >= 2,
+        InvalidLokiLabelsSnafu {
+            msg: "labels string too short"
+        }
+    );
+    ensure!(
+        labels.starts_with("{"),
+        InvalidLokiLabelsSnafu {
+            msg: "missing `{` at the beginning"
+        }
+    );
+    ensure!(
+        labels.ends_with("}"),
+        InvalidLokiLabelsSnafu {
+            msg: "missing `}` at the end"
+        }
+    );
+
+    let mut result = BTreeMap::new();
+    labels = &labels[1..labels.len() - 1];
+
+    while !labels.is_empty() {
+        // parse key
+        let first_index = labels.find("=").context(InvalidLokiLabelsSnafu {
+            msg: format!("missing `=` near: {}", labels),
+        })?;
+        let key = &labels[..first_index];
+        labels = &labels[first_index + 1..];
+
+        // parse value
+        let qs = quoted_string::parse::<TestSpec>(labels)
+            .map_err(|_| {
```

Comment:
Inline the error into the final message. Otherwise it would be hard to debug when something goes wrong.